### ISSUE TITLE
Towards STEP importing

### ISSUE THREAD

**Issue by agordon** [openscad/openscad#3065]
Hello,

I've been experimenting with code to read STEP files, and have some working prototype (without openSCAD). It uses "openCASCADE" to do the actual reading (just like almost any other STEP-capable free software out there), and generates either STL or SCAD scripts.

Currently, the output is very similar to STL (i.e., the object is first triangulated, then OpenSCAD draws the triangles using `polyhedron`) - but there's some code to explore the class heirarchy  of OpenCASCADE too - if someone wants to give it a try.

A fun side-effect of reading the STEPs is being able to seperate the planes, and give each one a different color:

![32mmChannel](https://github.com/agordon/openscad-step-reader/blob/master/examples/tetrix-32mmChannel/channel-faces.png?raw=true])

Code and examples here https://github.com/agordon/openscad-step-reader .

Comments and feedback are very welcomed.

gordon

**Comment by t-paul** [comment]
Very cool. OpenCASCADE is a bit scary though as we already have quite some trouble getting the CI builds into the allowed time frames. I was looking at stepCODE for exporting STEP but did not have enough time to really dig into it. Having OpenCASCADE available would be nice also from perspective of maybe allowing it as alternative CSG engine.

I tried compiling the reader but got some linker error hinting at some c++abi incompatibility, not yet sure where that's coming from. I'm on Debian/testing using gcc9 which should not be much different yet from Buster mentioned in the build notes.

**Comment by agordon** [comment]
I agree OpenCASCADE is a PITA.
That's why I was thinking of keeping it as a separate executable, which can be optionally built if the needed libraries are there, or skipped other wise.

Later, we can add OpenSCAD menu option to import/convert by executing the binariy. If it's missing - not a big problem - just show error and explain.

**Comment by kintel** [comment]
I like the idea of an executable converting files to .scad (or .csg) format, and possibly some UI feature to insert (or reference using import/include) in the editor.

If we generalize this, we could do the same for other 2D and 3D file formats, which could be useful as well.

That said, calling out into a separate process is a new feature that may create new challenges, but the only way to find out is to try I guess ;)

**Comment by jack9603301** [comment]
That's a good idea. When can we support it?

**Comment by t-paul** [comment]
Yep, the idea to start decoupled from the main application is great as the topic is complicated enough on it's own. But unfortunately I'm not aware of anyone working on this.

**Comment by tpimh** [comment]
There's a new library for working with STEP: [ruststep](https://github.com/ricosjp/ruststep) (yes, it's being written in Rust)

**Comment by tcurdt** [comment]
unstale - just because there is bot now does not change the fact why it was opened

**Comment by gsohler** [comment]
this is a  a very good reference to learn, how to import step without HUGE external libraries :   

https://github.com/slugdev/stltostp/

I was using it.

**Comment by kintel** [comment]
This isn't really an OpenSCAD issue, it's more of a discussion and should be moved to the mailing list.
